### ISSUE TITLE
`keyof` on homomorphic mapped type over an array thinks keys are function types

### ISSUE THREAD

**Issue by TotallyNotChase** [microsoft/TypeScript#44217]
# Bug Report

Creating a mapped type (where each value is just the key - `{ [K in keyof T]: K }`) when applied over a generic type `T`, where `T` has been passed as an array type, somehow manages to resolve the keys of an array as functions. I've read the FAQ about homomorphic mapped types but I can't imagine the `K` in `K in keyof T` ending up as a function type - no matter what the `T` eventually becomes.

### 🔎 Search Terms

keyof, array, types, type constructors

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about homomorphic mapped types

Tested on 3.9.7, 4.2.3.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=3.9.7#code/C4TwDgpgBA0hIGcA8AVAfFAvFA3gWACgpioBtGKASwDsoBreAewDMoUBdALlgG5CBfUgxAs27PgUKhIUAOIosseMgTAATjQDmpdmgkB6fVAB6UABKMAbhDVQAFjYgA6JYigBDNdGYBXagGNgSkZqBCgAfkjCIA)

### 💻 Code

```ts
type Keys<T> = {
    [K in keyof T]: K;
}[keyof T];

type GT = Keys<string[]>;
// ^ Hover here. Keys are functions ???
```

![Type upon hovering](https://user-images.githubusercontent.com/44284917/119232164-cd813f00-bb41-11eb-8db8-b47d4863ad4f.png)

### 🙁 Actual behavior

`Keys` applied over an array, returns a union of functions.

### 🙂 Expected behavior

`Keys` applied over an array, should return the property names of the array.

**Comment by fatcerberus** [comment]
This happens because mapping over `string[]` in the way you did produces `number[]`, which you then index with `keyof T`.  Since T is `string[]`, "every valid key of `T`" also includes the names of array methods such as `toString` and `concat`, which you can see if you take the result of `keyof T`.  This is why you end up with functions.

Still seems like it might be a bug since the mapped type shouldn't produce `number[]`, as you've replaced all the values with their keys.  `number[]` doesn't have `concat: "concat"` 😛  That said, your type--even if it worked properly--is functionally equivalent to `keyof T`, which means you'd get all the method names. Probably not very useful in case of an array type.

**Comment by TotallyNotChase** [comment]
@fatcerberus Yeah. I'd understand if it resolved to `number[]` if the mapped type was `[K in keyof T]: T[K]`, but this just doesn't make sense.

Also the example here is just a tiny reproducible example. The actual mapped type I'd use over a generic T is much more complex.

**Comment by fatcerberus** [comment]
I think there might be a special case involved where mapping over an array type *only* maps the array part (i.e. the part indexable by `number`), which makes sense, but that makes it a bit confusing that `keyof T[]` then also includes the prototype props.

We’ll need to wait for a TS maintainer to chime in, I guess.

**Comment by RyanCavanaugh** [comment]
Something has gone terribly wrong.

Anyway a good workaround that probably (?) does what you wanted to happen anyway is:
```ts
type Keys<T> = {
    [K in keyof T & string]: K;
}[keyof T & string];
```
